src/Common/src/CoreLib/System/IO/PathInternal.cs

```diff
@@ -117,24 +117,54 @@ internal static bool AreRootsEqual(string first, string second, StringComparison
         internal static string RemoveRelativeSegments(string path, int rootLength)
         {
             Debug.Assert(rootLength > 0);
-            bool flippedSeparator = false;
+            
+            Span<char> initialBuffer = stackalloc char[260 /* PathInternal.MaxShortPath */];
+            ValueStringBuilder sb = new ValueStringBuilder(initialBuffer);
+
+            bool flippedSeparator = RemoveRelativeSegmentsHelper(path.AsSpan(), rootLength, ref sb);
+
+            // If we haven't changed the source path, return the original
+            if (!flippedSeparator && sb.Length == path.Length)
+            {
+                sb.Dispose();
+                return path;
+            }
+
+            return sb.Length < rootLength ? path.Substring(0, rootLength) : sb.ToString();
+        }
+
+        /// <summary>
+        /// Try to remove relative segments from the given path (without combining with a root).
+        /// </summary>
+        /// <param name="rootLength">The length of the root of the given path</param>
+        internal static void RemoveRelativeSegments(ReadOnlySpan<char> path, int rootLength, ref ValueStringBuilder sb)
+        {
+            Debug.Assert(rootLength > 0);
+            bool flippedSeparator = RemoveRelativeSegmentsHelper(path, rootLength, ref sb);
+
+            if (sb.Length < rootLength)
+            {
+                sb.Length = rootLength;
```

Comment:
I'm not understanding the logic here.  Couldn't there be garbage in the builder between sb.Length and rootLength?

Reply:
No, its not possible to have garbage in that length, in ```RemoveRelativeSegmentHelper``` we are first adding the characters the the builder upto the skip length (rootlength or rootlength  - 1 ) and then we never manipulate this part of array ever after.